[ResponseOps] [Cases] Add fields to the Cases SO for total alerts and comments.

adcoelho opened the pull request:
Fixes #147992

## Summary

- Added two new fields to the Cases' Saved Object:
  - `total_alerts`
  - `total_comment`
- Both fields are of the type _integer_.
- The default value for both fields is -1.
- Added tests for the migrations.
- Updated the `CaseAttributesRt` type with these two new fields. 
- **No other changes in the BE or FE, we anticipated the need for these fields and are doing the migrations in advance.**

Note that the total comments field is called `total_comment`, this was discussed offline and done to keep in line with the field's name already returned when fetching cases (`totalComment`).

This is a nitpick but I think I would instead change it to `total_comments`. What do you guys think?

adcoelho committed:
Add two new fields to the Cases SO, total_alerts and total_comment.
Tests.

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

elasticmachine commented:
Pinging @elastic/response-ops-cases (Feature:Cases)

adcoelho committed:
Updated the cases snapshot.

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into total-alerts-total-comment-migrations

adcoelho commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into total-alerts-total-comment-migrations

peteharverson commented on x-pack/plugins/cases/server/saved_object_types/cases.ts:
```diff
@@ -179,6 +179,12 @@ export const createCaseSavedObjectType = (
+      total_alerts: {
+        type: 'integer',
+      },
+      total_comment: {
```
Is it a big job to change this to `total_comments`? I agree with your suggestion in the description of the PR that `total_comments` is a more logical name than `total_comment`, and would be consistent with the plural form used for `total_alerts`.

adcoelho replied:
It is trivial to change!

Initially, @cnasikas suggested keeping the names consistent with the existing ones, but I prefer `total_comments`.

Maybe there is something else I am missing?

cnasikas replied:
Yes, because the API returns `totalComment` for each case I suggested keeping it consistent. I agree `total_comments` is better. When we use it, we must remember to map it to `totalComment`.

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into total-alerts-total-comment-migrations

adcoelho committed:
Added default value for totalAlerts in security solution.

jamster10 reviewed (approved):
Reviewed on behalf of @elastic/security-threat-hunting-explore. Looks good!

rudolf reviewed (commented):
This PR adds mappings which aren't searched/aggregated over. If there's no immediate plans to query these fields they should not be added to the mappings.

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into total-alerts-total-comment-migrations

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into total-alerts-total-comment-migrations

adcoelho committed:
Rename field to total_comments.
Move fields in type declaration to ESobject.

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into total-alerts-total-comment-migrations

jonathan-buttner reviewed (changes requested):
The mapping changes look good but I think we're going to need some backend and maybe frontend changes before we can merge this.

I tried importing an older security solution case from here: https://github.com/elastic/kibana/blob/main/x-pack/test/cases_api_integration/common/fixtures/saved_object_exports/single_case_user_actions_one_comment.ndjson

Which includes a comment. When you go to the Security Solution page it shows the total alerts at -1

![image](https://user-images.githubusercontent.com/56361221/211654306-1dc20d1b-098c-411a-86c2-005f9c6a04e4.png)

I believe this is happening because of how we convert snake case to camel case in the frontend

![image](https://user-images.githubusercontent.com/56361221/211654873-a60d69d1-2c55-42b6-9f38-51b61be4355e.png)


And the `totalAlerts` gets overwritten with the `total_alerts: -1` field. It's also concerning that our backend and frontend do not throw when there are excess fields returned, but that's unrelated to your changes.

I think the backend will need to remove these fields before returning them to the frontend until we're ready to use them.

The other issue I think is present is that new cases will not have the `total_*` fields set to -1. We're only setting those fields for the migrated cases. I think we need to set those fields to -1 when a case is created as we're doing in the migrations.

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into total-alerts-total-comment-migrations

cnasikas reviewed (commented):
I agree with Jonathan. Until we actually use the new fields we should not return them. The `CaseService` should remove them in the `transform<*>ResponseToExternalModel` Also, the fields should be part of the `ESCaseAttributes` and not the `CaseAttributes`. Another approach is not to migrate the fields at all. Cases without the fields are old cases and cases with the fields are new cases. Both approaches are fine (migrating to `-1` and not migrating at all). The second approach removes the necessity to remove the fields on the `CaseService`.

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into total-alerts-total-comment-migrations

adcoelho committed:
Addressing PR comments.
Fixed Cases creation.
Fixed services return value.

adcoelho commented:
Thank you guys for the review!

@jonathan-buttner 

1 - I reproduced the issue with the imported file and fixed it.  `transformSavedObjectToExternalModel` no longer returns the two new fields so the `snakeToCamel` conversion does not happen and the values are not overwritten with -1.
2 - I changed the creation to include the default values for the new fields. I did this in `postNewCase` and not in `transformAttributesToESModel` because that transformation is also used for `patchCase` and `patchCases` and I do those fields to be touched if not necessary.
3 - I updated the tests in `x-pack/plugins/cases/server/services/cases/index.test.ts` to take all this into account.

@cnasikas

> Also, the fields should be part of the ESCaseAttributes and not the CaseAttributes

Initially, things were not like this but after we talked I fixed it. This was before your comment though so maybe I forgot some other place? See https://github.com/elastic/kibana/pull/148202/commits/e0cad5c31f65398239dfa60bdba8bb266353e71c.

@rudolf 

> This PR adds mappings which aren't searched/aggregated over. If there's no immediate plans to query these fields they should not be added to the mappings.

That was actually on purpose. We are doing this before the migration freeze so that we can work on sorting from these fields as soon as 8.7 is done.

jonathan-buttner reviewed (commented):
Thanks for making those changes @adcoelho I left a few more comments.

cnasikas commented on x-pack/plugins/cases/server/services/cases/index.ts:
```diff
@@ -564,11 +564,16 @@ export class CasesService {
       this.log.debug(`Attempting to POST a new case`);
       const transformedAttributes = transformAttributesToESModel(attributes);
+
+      transformedAttributes.attributes.total_alerts = -1;
```
I think we should be consistent as to where we are defying the variables even if it is not perfect. `transformNewCase` seems more appropriate as is where we define all the other attributes. It seems that the `postNewCase`'s argument definition forces consumers (the case client for example) to think about the structure in ES and provide the data accordingly. In the future, it may be nice to refactor and move the logic of the `transformNewCase` inside the `postNewCase` function so consumers will only pass `CasePostRequest` objects.

adcoelho replied:
`transformNewCase` might make sense but in the future, because right now the return type is `CaseAttributes` which does not have `total_alerts` or `total_comments`. These fields are internal so that is why I thought it would make sense to have them closer to the transformation to ESModel.

> It seems that the postNewCase's argument definition forces consumers (the case client for example) to think about the structure in ES and provide the data accordingly.

But right now we don't want them to think about anything. The fields will quietly be there until there is an equivalent in `CaseAttributes`.

cnasikas replied:
I missed that the `transformNewCase` returns a `CaseAttributes` object which does not include `total_alerts` and `total_comments` as they are `ESCaseAttributes`. Make sense to me then. Also, because `transformAttributesToESModel` is also used for the update we do not want to put it there, correct? @jonathan-buttner What do you think?

jonathan-buttner replied:
Yeah good points, leaving it here is fine.

cnasikas commented:
> Initially, things were not like this but after we talked I fixed it. This was before your comment though so maybe I forgot some other place? See https://github.com/elastic/kibana/commit/e0cad5c31f65398239dfa60bdba8bb266353e71c.

Thanks, @adcoelho! No, I missed it while reviewing.

cnasikas commented on x-pack/plugins/cases/server/services/cases/transform.ts:
```diff
@@ -192,6 +192,11 @@ export function transformSavedObjectToExternalModel(
     STATUS_ESMODEL_TO_EXTERNAL[caseSavedObject.attributes?.status] ?? CaseStatuses.open;
 
+  // @ts-ignore
+  delete caseSavedObject.attributes?.total_alerts;
```
Let's not mutate the `caseSavedObject` and spread to remove the fields as @jonathan-buttner suggested.

cnasikas commented on x-pack/plugins/cases/server/services/cases/transform.ts:
```diff
@@ -192,6 +192,11 @@ export function transformSavedObjectToExternalModel(
+  // @ts-ignore
+  delete caseSavedObject.attributes?.total_alerts;
+  // @ts-ignore
+  delete caseSavedObject.attributes?.total_comments;
```
Shouldn't we do the same for the `transformUpdateResponseToExternalModel`?

adcoelho replied:
Yup, just changed it. I had added tests for `patch` but there was a bug in them. I fixed them, added one for `bulkPatch` too, and changed `transformUpdateResponseToExternalModel` 👍

cnasikas replied:
Amazing! I was wondering why the tests did not catch it 🙂.

adcoelho commented on x-pack/plugins/cases/server/services/cases/transform.ts:
```diff
@@ -192,6 +192,11 @@ export function transformSavedObjectToExternalModel(
   const status =
     STATUS_ESMODEL_TO_EXTERNAL[caseSavedObject.attributes?.status] ?? CaseStatuses.open;
 
+  // @ts-ignore
```
I had it like that initially but it would fail if `total_alerts` and `total_comments` weren't there so instead of an ugly chain of `if`s to check the properties I picked the `delete`s.

But `?? { total_alerts: -1, total_comments: -1, } as ESCaseAttributes;` is a better solution, I'll do it 😉

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into total-alerts-total-comment-migrations

cnasikas commented on x-pack/plugins/cases/server/services/cases/transform.ts:
```diff
@@ -192,6 +192,11 @@ export function transformSavedObjectToExternalModel(
   const status =
     STATUS_ESMODEL_TO_EXTERNAL[caseSavedObject.attributes?.status] ?? CaseStatuses.open;
 
+  // @ts-ignore
```
@jonathan-buttner How do you get this type of error? I am not able to reproduce it. I remember discussing this with @adcoelho and it seems that `myObject[undefined] === undefined` which will fallback to the expression after the `??`.

adcoelho replied:
> I remember discussing this with @adcoelho and it seems that myObject[undefined] === undefined which will fallback to the expression after the ??.

Yup that was it.

So if like you said @jonathan-buttner 

> the attributes field will be undefined. For the status and severity I think this might cause an indexing error:

and we do `SEVERITY_ES_MODEL_TO_EXTERNAL[undefined]` the value returned should be `CaseSeverity.LOW`.

jonathan-buttner replied:
> @jonathan-buttner How do you get this type of error? I am not able to reproduce it. I remember discussing this with @adcoelho and it seems that `myObject[undefined] === undefined` which will fallback to the expression after the `??`.
Ah I didn't realize that. We would get an indexing error if we typed the `caseSavedObject` to have an optional `attributes` field which is technically possible if a bulk response encounters an error for one of the ids. It doesn't need to be in this PR but I think we should change the type eventually otherwise it looks unnecessary as to why we have those `?` and it'd be easy to forget one in the future.

adcoelho committed:
Fix patch test
Add bulkPatch test
Update transformUpdateResponseToExternalModel

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into total-alerts-total-comment-migrations

adcoelho committed:
Update Cases snapshot

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into total-alerts-total-comment-migrations

jonathan-buttner reviewed (approved):
Tested and looks good! Thanks for addressing the feedback!

js-jankisalvi reviewed (approved):
Tested again and looks good!! Great work 🎉
